### PR TITLE
[SPARK-25951][SQL] Ignore aliases for distributions and orderings

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

When we canonicalize an `Expression`, we do not remove `Alias`. So two expressions which are the same but are renamed are considered to be semantically different. As we rely on semantic equality in order to check if the result is the same for 2 expressions, some optimizations - as showed in the JIRA - may fail to apply, eg. removing redundant shuffles, when a column is renamed.

The PR proposes to ignore `Alias`es when checking whether distributions and orderings are satisfied by introducing a new method `sameResults` which ignore `Alias`es.

Credit should be given to @maropu for the approach suggestion which follows #17400.

Closes #17400.

## How was this patch tested?

added UT

### CODE REVIEW

**Comment by rxin** [comment]
i didn't look at your new code, but is your old code safe? e.g. a project that depends on the new alias.

**Comment by mgaido91** [comment]
cc @cloud-fan @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -195,14 +195,26 @@ abstract class Expression extends TreeNode[Expression] {
...
     deterministic && other.deterministic && canonicalized == other.canonicalized
 
+  /**
+   * Returns true when two expressions will always compute the same result, even if the output may
+   * be different, because of different names or similar differences.
+   * Usually this means they their canonicalized form equals, but it may also not be the case, as
+   * different output expressions can evaluate to the same result as well (eg. when an expression
+   * is aliased).
+   */
+  def sameResult(other: Expression): Boolean = other match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I know it's always safer to introduce a new API, does is it really necessary? In `Canonicalize`, we erase the name for attributes, I think it's reasonable to erase the name of `Alias`, as it doesn't affect the output.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: "erase the name" can also mean remove `Alias`. If we can't clearly tell the difference between `semanticEquals` and `sameResult`, and give a guideline about using which one in which case, I think we should just update `semanticEquals`(i.e. `Canonicalize`).

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: remove `Alias` is not possible for the reason explained in https://github.com/apache/spark/pull/22957#issuecomment-436992955. In general, `semanticEquals` should be used when we want to replace an expression with another, while `sameResult` should be used in order to check that 2 expressions return the same output.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: can you put it in the method doc(both `semanticEquals` and `sameResult`)? This makes sense to me.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Sure, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -195,14 +195,26 @@ abstract class Expression extends TreeNode[Expression] {
...
+    case a: Alias => sameResult(a.child)
+    case _ => this.semanticEquals(other)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: can we also strip the alias of this here? so that we can mark `sameResult` as final.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I think it is doable, but I didn't want to put too many `match` where it was not needed. But if you prefer that way I can try and do that.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: well, it needs to be overridden by `HashPartitioning` too, so since I am not able to make it final anyway, I don't think it is a good idea. Well, I can add a match on `HashPartitioning`too, but it doesn't seem a clean solution to me.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: we can do
```
CleanupAliases.trimAliases(this) semanticEquals CleanupAliases.trimAliases(other)
```

**Comment by cloud-fan** [comment]
retest this please

**Comment by mgaido91** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -2542,10 +2542,10 @@ object EliminateUnions extends Rule[LogicalPlan] {
  * rule can't work for those parameters.
  */
 object CleanupAliases extends Rule[LogicalPlan] {
-  private def trimAliases(e: Expression): Expression = {
+  private[catalyst] def trimAliases(e: Expression): Expression = {
     e.transformDown {
-      case Alias(child, _) => child
-      case MultiAlias(child, _) => child
+      case Alias(child, _) => trimAliases(child)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: what's going on here?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: the point is that now this method removes only the first `Alias` it finds (and it doesn't go on recursively), which is the reason of the UT failure. Also checking the comment on the method it seems not the expected behavior of this method.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ah, I did a stupid thing here. So the problem is that: since it returns `child` for `this`, in transformDown we apply the rule to `child` children, instead of applying to `child` itself. So the problem here is with 2 consecutive `Alias`. Let me find a better fix.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: just using `transformUp`solves the issue

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 4]
```diff
@@ -780,6 +780,23 @@ class PlannerSuite extends SharedSQLContext {
         classOf[PartitioningCollection])
     }
   }
+
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: can we have an end-to-end test as well?

**Comment by cloud-fan** [comment]
LGTM, cc @viirya as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 5]
```diff
@@ -195,14 +195,35 @@ abstract class Expression extends TreeNode[Expression] {
...
+   * equivalent expressions). It should not be used (and `sameResult` should be used instead) when
+   * comparing if 2 expressions produce the same results (in this case `semanticEquals` can be too
+   * strict).
    */
   def semanticEquals(other: Expression): Boolean =
     deterministic && other.deterministic && canonicalized == other.canonicalized
 
+  /**
+   * Returns true when two expressions will always compute the same result, even if the output may
+   * be different, because of different names or similar differences.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I think here `output` is a bit confusing. Do we mean the output names?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: So sameResult returns if the evaluated results between two expressions are exactly the same?

**Comment by viirya** [comment]
This looks good to me. Just a comment about wording.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 4 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: +1 if possible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 5 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: yes, I mean: `sameResult` returns true if 2 expressions return the same data even though from plan perspective they are not the same (eg. the output name/exprIds is different as in this case), while `semanticEquals` ensure they are the same from plan perspective too. If you have better suggestions how to rephrase this, I am happy to improve it. Thanks.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: How about replace `output` with `output from plan perspective`?

```
Returns true when two expressions will always compute the same result, even if the output
from plan perspective may be different, because of different names or similar differences.
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 4 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: ah, good point and indeed very useful. In my previous tests I always used a very simple query to verify this and never the one reported in the JIRA. Now I tried that one and I realized that this fix is not very useful as of now, because in renaming like that in the `HashPatitioning` there is the `AttributeReference` to the `Alias`, rather than the `Alias` itself. Since that is the common case, the PR as it is now it is not very useful. If I won't be able to figure out a good way for that, I am going to close this. Thanks and sorry for the trouble.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: @cloud-fan @viirya I added the test, but as I mentioned I had to do another change in order to make it working. Sorry for the mistake. I'd really appreciate if you could review it again. Thanks.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: I think the test case coverage is not enough for such massive changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 6]
```diff
@@ -223,14 +223,35 @@ abstract class Expression extends TreeNode[Expression] {
...
+   * Usually this means that their canonicalized form equals, but it may also not be the case, as
+   * different output expressions can evaluate to the same result as well (eg. when an expression
+   * is aliased).
+   *
+   * This method should be used (instead of `semanticEquals`) when checking if 2 expressions
+   * produce the same results (eg. as in the case we are interested to check if the ordering is the
+   * same). It should not be used (and `semanticEquals` should be used instead) when comparing if 2
+   * expressions are the same and one can replace the other.
+   */
+  final def sameResult(other: Expression): Boolean =
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Add unit test cases to SubexpressionEliminationSuite?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: thanks, added

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 4 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: thanks @gatorsmile, I added also a negative case, but I don't think it is enough. Do you have some hints on cases to test? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 7]
```diff
@@ -145,9 +145,14 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     assert(requiredChildDistributions.length == children.length)
     assert(requiredChildOrderings.length == children.length)
 
+    val aliasMap = AttributeMap[Expression](children.flatMap(_.expressions.collect {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: is it safe to do so? I think we should only collect aliases from operators with a project list, i.e. `Project`, `Aggregate`, `Window`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 8]
```diff
@@ -145,9 +145,14 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
...
+      case a: Alias => (a.toAttribute, a)
+    }))
+
     // Ensure that the operator's children satisfy their output distribution requirements.
     children = children.zip(requiredChildDistributions).map {
-      case (child, distribution) if child.outputPartitioning.satisfies(distribution) =>
+      case (child, distribution) if child.outputPartitioning.satisfies(
+          distribution.mapExpressions(replaceAlias(_, aliasMap))) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: instead of doing it here, shall we deal with alias in `SparkPlan.outputPartitioning` for operators with a project list?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 7 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: I think it is. We are only checking the presence of aliases. In particular, we are collecting all the aliases which are defined in the previous operator. The solution you are suggesting works too IMHO and restricts the scope, but I am not sure it is a good thing, because I see no harm in doing it for other operators: simply they won't contain aliases; while I do see some issues in the maintenance of the "whitelist" of operators you are suggesting (we may miss some now or forget to update later...)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 8 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: this is not dealing with the aliases in the `outputPartitioning` but with the ones in the `requiredChildDistribution`. Anyway, I wouldn't do it there, because this would mean moving also the logic for collecting the aliases from the children there, which seems to me an operations which belong to a rule/transforming operator, rather than to the plan operator itself (eg. now these methods are in `PredicateHelper`...).

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: As an example, `ProjectExec.outputPartitioning` can be wrong, as it doesn't consider the aliases in the project list. I think it's clearer to adjust the `outputPartitioning` there, instead of dealing with it in a rule. What if we have more rules need to check `outputPartitioning` and `requiredChildDistribution`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: for example, `relation[a, b]`'s output partitioning is `[hash partition a, hash partition b]`, and `Project(a as c, b, relation)`'s output partitioning should be `[hash partition c, hash partition b]`.

What do you mean by `But ProjectExec.outputPartitioning cannot contain a reference to the aliases`?

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: ah I see now what you mean. I am not sure what you are suggesting is feasible. Imagine that in your example the Project is: `Project(a as c, a as d, b, relation)`. What should the output partitioning be?

> What do you mean by ...

I meant that when we collect the alias for `a as c`, we are mapping all the attr references of `c` with `a as c` here. In the `outputPartitioning`, there will never be an occurrence of a reference to `c`, but only references to `a`.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: For `Project(a as c, a as d, b, relation)`, I think the `outputPartitioning` should be `[hash part c, hash part d, hash part b]`. The point is, we should not report an output partitioning whose attribute is not even in the current plan's output.

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: I don't think that is right: that would cause the shuffle to happen for every plan which is hashed by both `[hash part c, hash part b]` and `[hash part d, hash part b]` (and also `[hash part a, hash part b]`). I think that if we want to go that way, we'd need a set of equivalent outputPatitioning

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: seems we are not on the same page...

Let's make the example clearer. Assuming a `relation[a ,b]`'s partitioning is `hash(a, b)`, then `Project(a as c, a as d, b, relation)`'s partitioning should be `[hash(c, b), hash(d, b)]`. It's like a flatMap.

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Ok, I think I got it now, sorry, I didn't understand :) yes I think this is doable then. Let me try and do that, thanks.

**Comment by mgaido91** [comment]
@cloud-fan @gatorsmile I updated the PR according to the previous suggestions and added a new dedicated test suite.
May you please review this again? Thanks.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
@cloud-fan @gatorsmile any more comments on this? Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 9]
```diff
@@ -1549,6 +1551,38 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       Seq(Row("Amsterdam")))
   }
 
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: I'm a little worried about introducing such a big change for such a corner case.

Looking at this test case, it seems updating the `outputPartitioning` according to the project list is good enough?

cc @maryannxue

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: > Looking at this test case, it seems updating the outputPartitioning according to the project list is good enough?

Basically, that is what is done. The diff is quite big, that's true, but mostly are tests. Is there something specific you are worried about?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: I'm worried about `Expression.sameResult`. Is it still useful after updating the outputPartitioning according to the project list?

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: it is still useful because it may happen a used writes something like:
```
select ... group by a as b, ...
```
(please notice that despite it makes no sense written in SQL, maybe with other APIs is easier to happen).

And especially it is useful when updating the `outputPartitioning` in order to check whether an expression produces the same result of another.

**Comment by mgaido91** [comment]
anymore comments on this?

**Comment by cloud-fan** [comment]
I don't really agree with this approach. I think we should just propagate partitioning and ordering through Project/Aggregat/Window and any other operators that have project list.

**Comment by mgaido91** [comment]
@cloud-fan thanks for your comment. You mean we should run `updatePartitioningWithNewOutput` only on those operators or do you have something else in mind? Thanks.

**Comment by cloud-fan** [comment]
I'm not sure if all the operators have same logic of propagating child partitioning and ordering. In general we want every operator to propagate children's partitioning and ordering if possible.

**Comment by mgaido91** [comment]
@cloud-fan sorry, I am not sure I am 100% getting what you mean, may I please ask you to elaborate a bit more? Thanks.

**Comment by cloud-fan** [comment]
by default `SparkPlan.outputPartitioning` is `UnknownPartitioning`, this is safe but is sub-optimal for some operators like `ProjectExec`.

`ProjectExec.outputPartitioning` just returns `child.outputPartitioning`, which is better but still sub-optimal. If we have `a as c`, the partitioning info of column `c` is not propagated.

**Comment by cloud-fan** [comment]
And I think this is the root cause of the problem, we don't need `Expression.sameResult `

**Comment by mgaido91** [comment]
yes, that is what this PR does, it propagates the partitioning info for column `c`. So I don't understand what your concerns are about the current approach. In particular that is performed only for Aggregates, Project and Window (we can extend this list in the future if we see more operators can do the same).

**Comment by mgaido91** [comment]
ah, I see now your point. So let me elaborate it a bit more, probably I was not clear. There are 2 cases which are addressed here:
 1. `select * from t1 join (select a as b, count(*) from t2 group by a) on t1.b=t2.b`: this is the case we are referring to and the `sameResult` usage in the check for the Partitionings/Orderings is not needed;
 2. there is a weirder case, when the alias is used in the ORDER BY/GROUP BY clause (eg. `select ... group by a as b, ...`). For this case we need the `sameResult` there.

So there are 2 cases fixed in 2 different ways. If you prefer, we can focus only on the first case for this PR and remove the usage of `Expression.sameResult` in checking the Partionings/Orderings. What do you think?

**Comment by mgaido91** [comment]
let me try and do that, thanks

**Comment by mgaido91** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 10]
```diff
@@ -186,6 +187,69 @@ trait Partitioning {
...
+          case PartitioningCollection(partitionings) => partitionings
+          case other => Seq(other)
+        }
+        val validPartitionings = exprToEquiv.foldLeft(initValue) {
+          case (partitionings, (toReplace, equivalents)) =>
+            if (equivalents.isEmpty) {
+              partitionings.map {
+                case hp: HashPartitioning if hp.references.contains(toReplace) =>
+                  UnknownPartitioning(hp.numPartitions)
+                case rp: RangePartitioning if rp.references.contains(toReplace) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: since we need to handle different partitioning differently, shall we add a method to `Partitioning`?

**Comment by HeartSaVioR** [comment]
Sorry about jumping in, but looks like we missed #17400 (SPARK-19981) which looks like containing the fix what @cloud-fan suggests - reviewers just lost focus on that. For me #17400 looks very concise, so curious how this patch is different from #17400.

**Comment by mgaido91** [comment]
@HeartSaVioR thanks for pointing that out! Yes, that fix is similar to this, the main difference between these 2 PRs is that this one handles also the case when we have `select a as b, a, ...` and then we have the partitioning on `a`, while #17400 doesn't (and hence may introduce a regression). Let me cc @maropu too here, since he worked on this and may have comments/suggestion or he can update his PR to handle that case and we can focus on that PR.

**Comment by maropu** [comment]
@mgaido91 Thank for letting me know! Could you fix this pr based on my fix? If you can, its ok to close my pr in #17400.

**Comment by mgaido91** [comment]
@maropu thanks for checking this. Do you mean using the trait approach? If so, sure, I am doing. If not, please let me know. Thanks.

**Comment by HeartSaVioR** [comment]
Just a 2 cents, personally I understood @maropu's comment as taking up his PR, in other words, rebasing this branch to his branch to retain his commits, and adding your commit on top of his branch to fix up remaining issue. This would make giving authorship of this PR to both of @maropu and @mgaido91 easier.

**Comment by maropu** [comment]
Thanks, @mgaido91 and @HeartSaVioR! The fix @mgaido91 did looks ok to me. But, I don't review yet and I'll do in a few days.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, PartitioningCollection, UnknownPartitioning}
+
+/**
+ * Trait for plans which can produce an output partitioned by aliased attributes of their child.
+ * It rewrites the partitioning attributes of the child with the corresponding new ones which are
+ * exposed in the output of this plan. It can avoid the presence of redundant shuffles in queries
+ * caused by the rename of an attribute among the partitioning ones, eg.
+ *
+ * spark.range(10).selectExpr("id AS key", "0").repartition($"key").write.saveAsTable("df1")
+ * spark.range(10).selectExpr("id AS key", "0").repartition($"key").write.saveAsTable("df2")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala]: do you mean view here? otherwise the physical plan doesn't match.

**Comment by mgaido91** [comment]
@HeartSaVioR @maropu rebasing this PR with @maropu 's change and work on it is non-trivial, I can rather close this and create a new one based on @maropu 's branch if you prefer. Otherwise committers can add @maropu as author when this will be merged or @maropu can take this over and create a new PR based on this. Just let me know how you prefer to go on, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 12]
```diff
@@ -235,6 +241,14 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
    * than numPartitions) based on hashing expressions.
    */
   def partitionIdExpression: Expression = Pmod(new Murmur3Hash(expressions), Literal(numPartitions))
+
+  override private[spark] def pruneInvalidAttribute(invalidAttr: Attribute): Partitioning = {
+    if (this.references.contains(invalidAttr)) {
+      UnknownPartitioning(numPartitions)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: `HashPartitioning('a, 'b)` with output expressions `'a as 'a1`, should produce `UnknownPartitioning` or `HashPartitioning('a1)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 13]
```diff
@@ -284,6 +298,19 @@ case class RangePartitioning(ordering: Seq[SortOrder], numPartitions: Int)
       }
     }
   }
+
+  override private[spark] def pruneInvalidAttribute(invalidAttr: Attribute): Partitioning = {
+    if (this.references.contains(invalidAttr)) {
+      val validExprs = this.children.takeWhile(!_.references.contains(invalidAttr))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: `this.children` -> `ordering`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 14]
```diff
@@ -284,6 +298,19 @@ case class RangePartitioning(ordering: Seq[SortOrder], numPartitions: Int)
...
+      if (validExprs.isEmpty) {
+        UnknownPartitioning(numPartitions)
+      } else {
+        RangePartitioning(validExprs, numPartitions)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: think about `RangePartitioning('a.ASC, 'b.ASC)` with output expression `'a as 'a1`.

It cannot satisfy `ClusteredDistribution('a1)`, but can still satisfy `OrderedDistribution('a1.ASC)`. I think the expected result should be `RangePartitioning('a1.ASC, 'b.ASC)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+   */
+  protected def outputExpressions: Seq[NamedExpression]
+
+  /**
+   * Returns the valid `Partitioning`s for the node w.r.t its output and its expressions.
+   */
+  final override def outputPartitioning: Partitioning = {
+    child.outputPartitioning match {
+      case partitioning: Expression =>
+        val exprToEquiv = partitioning.references.map { attr =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala]: can you explain what's going on here? The code is a little hard to follow.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 14 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: Why doesn't it satisfy `ClusteredDistribution('a1)`? I don't agree with what you stated. If `b` is not in the output, it is useless to have it there. Moreover, when ordering we order always for the first attribute, then for the second, ... So if something is partitioned by `RangePartitioning('a1.ASC, 'b.ASC)`, it is also true that its partitioning is `RangePartitioning('a1.ASC, )`. So I think that in that case `RangePartitioning('a1.ASC)` is the right one.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala (hunk 15 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala]: sure, let me add some comments. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 14 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: according to `RangePartitioning.satisfies0`, and the comments in `RangePartitioning` and `ClusteredDistribution`, `RangePartitioning('a.ASC, 'b.ASC)` does not satisfy `ClusteredDistribution('a)`.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: mmmh, I am not sure about what you mean referring to the classdoc of these 2 classes. I see nothing there about this. Anyway, I see that the implementation is done according what you state, but I do believe that it is wrong (or at least suboptimal if you prefer). If the data is partitioned by sorting it with `a.ASC, b.ASC`, it is definitely partitioned by sorting it with `a.ASC`. I think that `forall` should be an `exists`. There is also a (very minor) bug in the current implementation; try and running this test (it fails...):
```
  test("partitioning test") {
    val attr1 = AttributeReference("attr1", IntegerType)()
    val attr2 = AttributeReference("attr2", IntegerType)()
    val partitioning = RangePartitioning(Seq.empty, 10)
    val requiredDistribution = ClusteredDistribution(Seq(attr2, attr1), Some(10))
    assert(!partitioning.satisfies(requiredDistribution))
  }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: please carefully read the classdoc of `RangePartitioning` and `ClusteredDistribution`, and see what `RangePartitioning` guarantees and what `ClusteredDistribution` requires.
```
partition 1: (a=1, b=2), (a=1, b=3)
partition 2: (a=1, b=4), (a=1, b=5)
```
This data set is range partitioned by a,b, but not clustered by a.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: ah I see now, thanks. Let me update it then accordingly, thanks.

**Comment by mgaido91** [comment]
@cloud-fan @maropu sorry for pinging you again, I think I addressed all your comments on this, may you please check it again? Thanks.
